hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/HederaStateImpl.java

```diff
@@ -0,0 +1,234 @@
+
+    // For serialization
+    private static final long CLASS_ID = 29399209029302L;
+    private static final int VERSION_0 = 0;
+    private static final int CURRENT_VERSION = VERSION_0;
+
+    private BiConsumer<Round, SwirldDualState> onHandleConsensusRound;
+
+    /**
+     * Create a new HederaState! This can be called either explicitly or as part of saved state
+     * loading.
+     */
+    public HederaStateImpl() {}
+
+    /**
+     * Private constructor for fast-copy.
+     *
+     * @param from The other state to fast-copy from. Cannot be null.
+     */
+    private HederaStateImpl(@Nonnull final HederaStateImpl from) {
+        // Copy the Merkle route from the source instance
+        super(from);
+
+        // Copy the non-null Merkle children from the source (should also be handled by super, TBH).
+        for (int childIndex = 0, n = from.getNumberOfChildren(); childIndex < n; childIndex++) {
+            final var childToCopy = from.getChild(childIndex);
+            if (childToCopy != null) {
+                setChild(childIndex, childToCopy.copy());
+            }
+        }
+
+        // **MOVE** over the listener
+        this.onHandleConsensusRound = from.onHandleConsensusRound;
+        from.onHandleConsensusRound = null;
+    }
+
+    @Override
+    public ReadableStates createReadableStates(@Nonnull final String serviceName) {
+        final var opt = getServiceStateNode(serviceName);
+        if (opt.isEmpty()) {
+            return null; // Null, or optional??
+        }
+
+        return new ReadableStates() {
+            @Nonnull
+            @Override
+            public <K, V, S extends ReadableState<K, V>> S get(@Nonnull String stateKey) {
+                final var service = opt.get();
+                final var merkleNode = service.find(stateKey);
+                if (merkleNode instanceof MerkleMap) {
+                    //noinspection unchecked,rawtypes
+                    return (S) new InMemoryState(stateKey, (MerkleMap) merkleNode);
+                } else if (merkleNode instanceof VirtualMap) {
+                    //noinspection unchecked,rawtypes
+                    return (S) new OnDiskState<>(stateKey, (VirtualMap) merkleNode);
+                } else {
+                    throw new IllegalArgumentException("Cannot find state for '" + stateKey + "'");
+                }
+            }
+        };
+    }
+
+    @Override
+    public WritableStates createWritableStates(@Nonnull final String serviceName) {
+        throwIfImmutable();
+        final var opt = getServiceStateNode(serviceName);
+        if (opt.isEmpty()) {
+            return null; // Null, or optional??
+        }
+
+        return new WritableStates() {
+            @Nonnull
+            @Override
+            public <K, V> WritableState<K, V> get(@Nonnull String stateKey) {
+                final var service = opt.get();
+                final var merkleNode = service.find(stateKey);
```

Comment:
If `createWritableStates()` is to be called once per-handle, I suspect we'll want to cache the `MerkleNode`'s found for the current mutable copy. (According to JProfiler, the `getChild()` calls don't get inlined, and eliminating *seemed* to have a notable performance impact.)

Reply:
Yes. This is a good idea.

Reply:
It turns out I now cache the WritableKVState / ReadableKVState for another reason related to migration correctness, so this is now resolved (i.e. we won't do any more lookups than necessary).